### PR TITLE
Environment fixes

### LINKED ISSUE #4780: Error message when virtual env folder already exists has replacement tag {0} in msg
image

### LINKED ISSUE #4782: Crash showing the prompt for create virtual env
Open VS , click "Create a new project", input "Flask" in the search input box, select "Flask Web Project" and then click next

Here is the error log: 
``` 
System.NullReferenceException: Object reference not set to an instance of an object. 
   at Microsoft.PythonTools.Project.VirtualEnvCreateInfoBar.ShowCreateVirtualEnvironment(String txtPath) 
   at Microsoft.PythonTools.Project.VirtualEnvCreateInfoBar.<CheckAsync>d__5.MoveNext() 
--- End of stack trace from previous location where exception was thrown --- 
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
   at Microsoft.PythonTools.Infrastructure.VSTaskExtensions.<HandleAllExceptions>d__5.MoveNext() 
--- End of stack trace from previous location where exception was thrown --- 
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject) 
```

### LINKED ISSUE #4786: Install packages info bar incorrectly determines packages are missing
Package name comparison should be case insensitive
Can be reproduced with django web project

### PR DESCRIPTION
**Pull request by huguesv**
Fix #4786 Install packages info bar incorrectly determines packages are missing

- Refactor the environments.txt helper code to a separate class, fix the case sensitive check.
- Refactor the tests and add a new test for the bug.

Fix #4782 Crash showing the prompt for create virtual env

- Refactor all 3 info bars to share code with a base class.
- If we fail to get the info bar host, which happens when the main window is not created yet, try again later when idle.
- Move the trigger to "file open" rather than "project open", that reduces the number of info bars that will appear if several projects are in the solution. They will appear only when one of the project files get opened.

Fix #4780 Error message when virtual env folder already exists has replacement tag {0} in msg

Also delete some code that was no longer being compiled but was left over on disk.

### CODE REVIEW

File: Python/Product/PythonTools/PythonTools/Project/PackageInstallInfoBar.cs [hunk 1]
```diff
@@ -127,38 +101,32 @@ internal sealed class PackageInstallInfoBar : IVsInfoBarUIEvents, IDisposable {
...
+            Create(new InfoBarModel(messages, actions, KnownMonikers.StatusInformation, isCloseButtonVisible: true));
         }
 
-        public void OnClosed(IVsInfoBarUIElement infoBarUIElement) {
-            infoBarUIElement.Unadvise(_adviseCookie);
-            _infoBar = null;
-        }
+        private async Task<bool> PackagesMissingAsync(IPackageManager pm, string txtPath) {
+            try {
+                var original = File.ReadAllLines(txtPath);
```
Comment (AlexanderSher) [inline Python/Product/PythonTools/PythonTools/Project/PackageInstallInfoBar.cs]: `ReadAllLinesAsync`

File: Python/Product/PythonTools/PythonTools/Project/PythonProjectInfoBar.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudioTools;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.PythonTools.Project {
+    internal abstract class PythonProjectInfoBar : IVsInfoBarUIEvents, IDisposable {
+        private readonly IVsShell _shell;
+        private readonly IVsInfoBarUIFactory _infoBarFactory;
+        private IdleManager _idleManager;
```
Comment (AlexanderSher) [inline Python/Product/PythonTools/PythonTools/Project/PythonProjectInfoBar.cs]: `readonly`?

File: Python/Product/PythonTools/PythonTools/Project/PythonProjectInfoBar.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            FinishCreate();
+        }
+
+        protected void Close() {
+            _infoBar.Close();
+        }
+
+        protected void FinishCreate() {
+            if (_infoBar != null) {
+                return;
```
Comment (AlexanderSher) [inline Python/Product/PythonTools/PythonTools/Project/PythonProjectInfoBar.cs]: Is it ok that `_infoBar` references `InfoBarModel` that is different from `_infoBarModel`?

File: Python/Product/PythonTools/PythonTools/Project/PythonProjectNode.cs [hunk 4]
```diff
@@ -2316,7 +2331,8 @@ bool elevate
             TaskDialog.CallWithRetry(
                 _ => {
                     if (items.Any()) {
-                        File.WriteAllLines(txt, MergeRequirements(existing, items, addNew));
+                        var merged = PipRequirementsUtils.MergeRequirements(existing, items, addNew);
+                        File.WriteAllLines(txt, merged);
```
Comment (AlexanderSher) [inline Python/Product/PythonTools/PythonTools/Project/PythonProjectNode.cs]: `WriteAllLinesAsync`

File: Python/Product/PythonTools/PythonTools/Project/PackageInstallInfoBar.cs (hunk 1 above)
Comment (huguesv) [reply Python/Product/PythonTools/PythonTools/Project/PackageInstallInfoBar.cs]: I don't think that exists?

File: Python/Product/PythonTools/PythonTools/Project/PythonProjectInfoBar.cs (hunk 3 above)
Comment (huguesv) [reply Python/Product/PythonTools/PythonTools/Project/PythonProjectInfoBar.cs]: It's extra protection that should never get it. If the info bar has been created, it's a programming error to want to create it again. In fact, I'll add an assert.

File: Python/Product/PythonTools/PythonTools/Project/PackageInstallInfoBar.cs (hunk 1 above)
Comment (AlexanderSher) [reply Python/Product/PythonTools/PythonTools/Project/PackageInstallInfoBar.cs]: Yeah, not in 4.7.2 :(

**Pull request merged**